### PR TITLE
session: preliminary implementation 1

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
 This is split from https://github.com/pmodels/mpich/pull/4949/commits to simplify review. The commits have been tested inside #4949 with session tests.

[skip warnings]
## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Passes whitespace checkers
* [ ] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

File: src/include/mpi.h.in [hunk 1]
```diff
@@ -872,10 +876,11 @@ typedef int (MPIX_Grequest_wait_function)(int, void **, double, MPI_Status *);
 #define MPI_T_ERR_INVALID_NAME      73  /* Name doesn't match */
 #define MPI_T_ERR_INVALID           74  /* Generic error code for MPI_T added in MPI-3.1 */
 
+#define MPI_ERR_SESSION             75  /* Invalid session handle */
```
Comment (hzhou) [inline src/include/mpi.h.in:879]: ref: https://github.com/pmodels/mpich/pull/4949/commits/128e87888eb40bd3ff6557d953a609dfd9a38acc#r559740084

Comments added

File: src/mpi/init/init_impl.c [hunk 2]
```diff
@@ -21,6 +21,171 @@
...
+    /* if the name buffer length is 0, just return needed length */
+    if (*pset_len == 0) {
+        *pset_len = len + 1;
+        goto fn_exit;
+    }
+
+    /* copy the name, trucate if necessary */
+    if (len > *pset_len - 1) {
+        len = *pset_len - 1;
+    }
```
Comment (hzhou) [inline src/mpi/init/init_impl.c:128]: ref: https://github.com/pmodels/mpich/pull/4949/commits/9ceb439918f39fb125d31b9e1229ae7d9c9ca103#r560347040

Comments added

File: src/mpi/init/init_impl.c [hunk 3]
```diff
@@ -21,6 +21,171 @@
...
+
+    mpi_errno = MPIR_Info_alloc(info_p_p);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    /* Currently we only support session in MPI_THREAD_MULTIPLE */
+    /* Multiple session may run in threads concurrently, so significant work is needed to support
+     * per-session MPI_THREAD_SINGLE */
+    /* TODO: support per-session MPI_THREAD_SINGLE and optimize */
+    mpi_errno = MPIR_Info_set_impl(*info_p_p, "mpi_thread_support_level", "MPI_THREAD_MULTIPLE");
+    MPIR_ERR_CHECK(mpi_errno);
```
Comment (hzhou) [inline src/mpi/init/init_impl.c:150]: ref: https://github.com/pmodels/mpich/pull/4949/commits/9ceb439918f39fb125d31b9e1229ae7d9c9ca103#r560351212

Explanation added

File: src/mpi/group/group_impl.c [hunk 4]
```diff
@@ -622,3 +622,47 @@ int MPIR_Group_union_impl(MPIR_Group * group_ptr1, MPIR_Group * group_ptr2,
...
+int MPIR_Group_from_session_pset_impl(MPIR_Session * session_ptr, const char *pset_name,
+                                      MPIR_Group ** new_group_ptr)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_Group *group_ptr;
+
+    if (strcmp(pset_name, "mpi://WORLD") == 0) {
+        mpi_errno = MPIR_Group_create(MPIR_Process.size, &group_ptr);
+        MPIR_ERR_CHECK(mpi_errno);
+
```
Comment (hzhou) [inline src/mpi/group/group_impl.c:635]: Ref: https://github.com/pmodels/mpich/pull/4949/commits/4de649bba2e1d2ba3294360bfefd3e547020ebba#r560627217

File: src/mpi/comm/comm_impl.c [hunk 5]
```diff
@@ -678,6 +678,72 @@ int MPIR_Comm_dup_with_info_impl(MPIR_Comm * comm_ptr, MPIR_Info * info, MPIR_Co
...
+     */
+
+    /* Need convert stringtag to tag, use macro from uthash.h to calculate hash value */
+    unsigned hash;
+    int n = strlen(stringtag);
+    HASH_VALUE(stringtag, n, hash);
+
+    /* NOTE: tag will be used with MPIR_TAG_COLL_BIT on, ref. MPIR_Get_contextid_sparse_group */
+    int tag = hash % (MPIR_Process.attrs.tag_ub);
+
```
Comment (hzhou) [inline src/mpi/comm/comm_impl.c:703]: ref: https://github.com/pmodels/mpich/pull/4949/commits/accb10b2756f1a55619eb2fd7b4fb6fdbf9df5fe

**Review by hzhou — COMMENTED** [review]
@minsii Here are the first 5 commits. Let's get this reviewed first.

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

File: src/mpi/init/init_impl.c (hunk 2 above)
Comment (minsii) [reply src/mpi/init/init_impl.c:141]: trucate-> truncate

File: src/mpi/comm/comm_impl.c (hunk 5 above)
Comment (minsii) [reply src/mpi/comm/comm_impl.c:707]: Respond to
> It is only two lines and it is only used here, making it a separate routine seems unnecessary. Do I need improve the comments?

The logic here is clear but just I feel that the logic is standalone, so might be more readable if we define a subroutine for it (not for reuse).

File: src/mpi/comm/comm_impl.c [hunk 6]
```diff
@@ -678,6 +678,72 @@ int MPIR_Comm_dup_with_info_impl(MPIR_Comm * comm_ptr, MPIR_Info * info, MPIR_Co
...
+    if (MPIR_Process.comm_world) {
+        if (!MPIR_Process.comm_world->local_group) {
+            /* need create local_group for MPIR_Comm_create_group_impl to work */
+            MPIR_Group *world_group_ptr;
+            mpi_errno = MPIR_Comm_group_impl(MPIR_Process.comm_world, &world_group_ptr);
+            MPIR_ERR_CHECK(mpi_errno);
+            MPIR_Group_release(world_group_ptr);
+        }
```
Comment (minsii) [inline src/mpi/comm/comm_impl.c:715]: The comment is OK. But `MPIR_Comm_group_impl ` is not really for this purpose. So I feel weird to use it here. Might be better if we define a specific routine to explicitly initialize `local_group`. We can use the routine here and  in `MPIR_Comm_group_impl`

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

File: src/mpi/comm/comm_impl.c [hunk 7]
```diff
@@ -711,42 +819,17 @@ int MPIR_Comm_get_name_impl(MPIR_Comm * comm_ptr, char *comm_name, int *resultle
...
 
     MPIR_FUNC_TERSE_ENTER(MPID_STATE_MPIR_COMM_GROUP_IMPL);
-    /* Create a group if necessary and populate it with the
-     * local process ids */
+    /* Create a local group if necessary */
     if (!comm_ptr->local_group) {
-        n = comm_ptr->local_size;
-        mpi_errno = MPIR_Group_create(n, group_ptr);
+        mpi_errno = comm_create_local_group(comm_ptr);
         MPIR_ERR_CHECK(mpi_errno);
```
Comment (minsii) [inline src/mpi/comm/comm_impl.c:833]: Refactoring of `comm_create_local_group ` should be in a separate commit

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

File: src/mpi/comm/comm_impl.c (hunk 7 above)
Comment (hzhou) [reply src/mpi/comm/comm_impl.c:833]: done

File: src/mpi/init/init_impl.c [hunk 8]
```diff
@@ -28,12 +29,164 @@
 
 MPIR_Session MPIR_Session_direct[MPIR_SESSION_PREALLOC];
 
-MPIR_Object_alloc_t MPIR_Session_mem = { 0, 0, 0, 0,
+MPIR_Object_alloc_t MPIR_Session_mem = { 0, 0, 0, 0, 0, 0,
     MPIR_SESSION, sizeof(MPIR_Session),
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpi/init/init_impl.c]: fixed

**Comment by hzhou** [comment]
@minsii PR #5016 has merged to `main` and this PR has been rebased. Could you do the final review?

**Review by minsii — APPROVED** [review]

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most
